[1.12] Migrate most of FluidRegistry to the ForgeRegistry system

Madsthunder opened the pull request:
This pr moves the registry related methods to the PersistentRegistry system while trying to maintain backwards compatibility for 1.11.

As always, with breaking changes, there will need to be some things changed on the modders side. This includes:
- Having to register a fluid through either `GameRegistry.register`, `ForgeRegistries.FLUIDS.register`, or subscribing a `RegistryEvent.Register<Fluid>` to the `EventBus`.
- Adding the corresponding modid to the beginning of a custom fluid model (See the bottom of the changed files).

Madsthunder committed:
Completely rewrite how fluids are registered

Madsthunder committed:
Completely rewrite how fluids are registered

Madsthunder committed:
Completely rewrite how fluids are registered

Madsthunder committed:
Completely rewrite how fluids are registered

Madsthunder committed:
Repurposed 'FluidName' tag as a fallback if no 'Fluid' tag is registered

Madsthunder committed:
Repurposed 'FluidName' tag as a fallback if no 'Fluid' tag is registered

Madsthunder committed:
Repurposed 'FluidName' tag as a fallback if no 'Fluid' tag is registered

Madsthunder committed:
Repurposed 'FluidName' tag as a fallback if no 'Fluid' tag is registered

Madsthunder committed:
Added missing Bootstrap patch

Madsthunder committed:
Added missing Bootstrap patch

Madsthunder committed:
Added missing Bootstrap patch

Madsthunder committed:
Added missing Bootstrap patch

Madsthunder committed:
Fixed the bootstrap patch (1)

Madsthunder committed:
Fixed the bootstrap patch (1)

Madsthunder committed:
Fixed the bootstrap patch (1)

Madsthunder committed:
Fixed the bootstrap patch (1)

liach commented on patches/minecraft/net/minecraft/init/Bootstrap.java.patch:
```diff
@@ -10,10 +10,11 @@
              StatList.func_151178_a();
              Biome.func_185358_q();
              func_151353_a();
++            net.minecraftforge.fluids.FluidRegistry.registerFluids();
```
I suggest making some method like `ForgeRegistries.initForgeRegistries` to prevent any future increase in patch size.

Madsthunder committed:
Add initRegistries method to ForgeRegistries

To prevent future patching of the Bootstrap class relating to registries

Madsthunder committed:
Add initRegistries method to ForgeRegistries

To prevent future patching of the Bootstrap class relating to registries

Madsthunder committed:
Add initRegistries method to ForgeRegistries

To prevent future patching of the Bootstrap class relating to registries

Madsthunder committed:
Add initRegistries method to ForgeRegistries

To prevent future patching of the Bootstrap class relating to registries

Madsthunder commented:
Should I split `FluidRegistry` into `Fluids` and `BucketRegistry`?

liach commented:
Not necessary.

HenryLoenwind commented:
I really don't like how this adds all the problems we have with the OreDictionary to the fluids, too. Fluids have had the most sensible dupe handling of everything: "There can only be one, first registered wins (for this game session)."

mezz commented:
Past fluid conflict resolution had its own issues actually. Client and server could disagree on who won and that would cause crashes, so mods like Forestry had to have configs for disabling fluids. Also it meant that fluids like BoP honey (a finite fluid) and Forestry honey (regular fluid) could not co-exist. I had to change the Forestry honey's fluid ID to `for.honey` which is not a great compromise. The `FluidDictionary.getFirstFluid` seems like an OK compromise.

Is there a way to auto-register with the fluid registry and the `FluidDictionary` though? I'm not sure it makes sense to have them separated for registration.

Madsthunder commented:
@mezz Should I un-deprecate the previous constructor and set it up to add the `Fluid` instance to the `FluidDictionary` automatically?

mezz commented:
Registering something from its constructor is pretty bad practice, I think what you have now is better.

The part that feels  clunky to me is this:
```java
GameRegistry.register(TestFluid.instance);
GameRegistry.register(TestGas.instance);
GameRegistry.register(milkFluid);
FluidDictionary.registerFluid(TestFluid.instance, TestFluid.name);
FluidDictionary.registerFluid(TestGas.instance, TestGas.name);
FluidDictionary.registerFluid(milkFluid, "milk");
```
But in the grand scheme of things it's not a really big deal.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,101 @@
+    
+    public static void registerFluid(Fluid fluid, String name)
+    {
+        name = name.toLowerCase();
```
Use `com.google.common.base.Preconditions#checkArgument` at the top of the method here to make sure the fluid is in the fluid registry first, it will prevent some confusion.

HenryLoenwind replied:
btw, toLowerCase without Locale is pure evil.

There are locales out there where ```"i" != "I".toLowerCase()```!

Madsthunder replied:
@HenryLoenwind Please explain?

HenryLoenwind replied:
The second line already was meant to be the explanation, so what's your question?

HenryLoenwind replied:
ok, example:

```
ModA: registerFluid(x, "moltenIron")
ModB: registerFluid(y, "molteniron")
```

In Sweden, this will give one dictionary entry. In Turkey it will give 2 different entries ("molteniron" and "moltenıron").

mezz replied:
To fix this issue, use `name = name.toLowerCase(Locale.ENGLISH);` when you are lowercasing unique IDs.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+    
+    public static NonNullList<String> getNames(Fluid fluid)
```
Everywhere you use this method it's `getNames(Fluid fluid).contains(String name)`.
You should replace this with `hasName(Fluid fluid, String name)` which just checks instead. 
That way you also avoid copying the list every time.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,101 @@
+    public static NonNullList<Fluid> getFluids(String name)
+    {
+        NonNullList<Fluid> ret = NonNullList.<Fluid>create();
+        if (NAME_TO_FLUIDS.containsKey(name = name.toLowerCase()))
```
Don't assign variables on the same line as something else, even if you know what you're doing. It's really bad style.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidStack.java:
```diff
@@ -84,17 +86,46 @@ public static FluidStack loadFluidStackFromNBT(NBTTagCompound nbt)
-
-        String fluidName = nbt.getString("FluidName");
-        if (FluidRegistry.getFluid(fluidName) == null)
+        //TODO Drop support for this in 1.14 or 1.15
```
Explain what this is for in the comment, so that future-whoever can decide if it's a good time to drop it.

Madsthunder replied:
@mezz Is `//TODO Drop legacy support for 'FluidNane' in 1.14 or 1.15, as it will be outdated` okay?

mezz replied:
That's better, yeah

mezz commented on src/main/java/net/minecraftforge/fluids/FluidRegistry.java:
```diff
@@ -281,14 +110,9 @@ public static boolean isUniversalBucketEnabled()
     public static boolean addBucketForFluid(Fluid fluid)
     {
-        if(fluid == null) {
+        if(fluid == null || !ForgeRegistries.FLUIDS.containsValue(fluid)) {
```
This `if` statement should be replaced with `com.google.common.base.Preconditions#checkNotNull` and `checkArgument`. The javadoc says it returns false if it was added already or couldn't be added, but realistically these two cases are completely invalid inputs and will be hard for devs to notice without a crash.

mezz commented on src/main/java/net/minecraftforge/fluids/BlockFluidBase.java:
```diff
@@ -150,7 +152,7 @@ public BlockFluidBase(Fluid fluid, Material material)
         this.disableStats();
 
-        this.fluidName = fluid.getName();
+        this.fluidName = fluid.getRegistryName();
```
Add `com.google.common.base.Preconditions#checkNotNull(fluid.getRegistryName())` at the top of this constructor because it's very easy to forget to set the registry name before calling this. It wasn't so easy to have a null `name` before because it was in the fluid constructor. This will help people port from the old registry to the new one more easily.

mezz replied:
Fixed above.

Madsthunder committed:
Apply some suggestions (1)

Madsthunder committed:
Apply some suggestions (1)

Madsthunder committed:
Apply some suggestions (1)

Madsthunder commented:
@mezz @HenryLoenwind I changed the references to `String#toLowerCase` to `String#getLowerCase(Locale)`, with the `Locale` being `Locale.ENGLISH`. I also added  `FluidDictionary#hasName` and added some `Preconditions` to `FluidRegistry`, `FluidDictionary`, and `BlockFluidBase`.

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -31,7 +34,8 @@
     public static void registerFluid(Fluid fluid, String name)
     {
-        name = name.toLowerCase();
+    	Preconditions.checkArgument(ForgeRegistries.FLUIDS.containsValue(fluid), "Cannot register FluidDictionary entry for unregistered fluid %s", fluid);
```
there's a tab here...

Madsthunder replied:
This is what I get for using Eclipse Oxygen...

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidStack.java:
```diff
@@ -100,7 +100,7 @@ public static FluidStack loadFluidStackFromNBT(NBTTagCompound nbt)
             }
         }
-        //TODO Drop support for this in 1.14 or 1.15
+        //TODO Drop legacy support for 'FluidNane' in 1.14 or 1.15, as it will be outdated 
```
"nane" => "name"

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/BlockFluidBase.java:
```diff
@@ -149,6 +149,7 @@
     public BlockFluidBase(Fluid fluid, Material material)
     {
         super(material);
+        Preconditions.checkNotNull(fluid.getRegistryName(), "Cannot register Fluid with null registry name %s", fluid);
```
"register fluid" => "create block for fluid"

HenryLoenwind commented on src/main/java/net/minecraftforge/client/model/ModelLoader.java:
```diff
@@ -391,10 +393,10 @@ public int compare(Item i1, Item i2)
             {
                 // can the milk be put into a bucket?
-                Fluid milk = FluidRegistry.getFluid("milk");
+                Fluid milk = FluidDictionary.getFirstFluid("milk");
```
What happens if only the second Milk can be put into buckets? (sry, honest question, I don't know the original code here...)

mezz replied:
All fluids can be put in the universal bucket unless it is disabled.
There are some odd comments in the original code, The check here seems to be seeing if a milk fluid is added, so I think the change here is valid.

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/Fluid.java:
```diff
@@ -126,10 +124,15 @@
     protected Block block = null;
 
+    //TODO Remove in 1.13
+    @Deprecated //Use constructor below
```
I don't quite see the reason in keeping this constructor---code that uses it must be changed anyway (setRegistryName+register)...

mezz replied:
That's a really good point, it's worth removing it now since everyone will have to fix their fluid code anyway.

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,114 @@
+    public static boolean nameExists(String name)
+    {
+        return NAME_TO_FLUIDS.containsKey(name.toLowerCase());
+    }
```
missed one

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,114 @@
+    public static FluidStack createFluidStack(String name, int amount)
+    {
+        Fluid fluid = getFirstFluid(name.toLowerCase());
+        return fluid == null ? null : new FluidStack(fluid, amount);
```
missed two

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidStack.java:
```diff
@@ -84,17 +86,46 @@ public static FluidStack loadFluidStackFromNBT(NBTTagCompound nbt)
+        else if (nbt.hasKey("FluidName", Constants.NBT.TAG_STRING))
+        {
+            String fluidName = nbt.getString("FluidName").toLowerCase();
+            Fluid fluid = null;
```
missed three

HenryLoenwind commented on src/main/java/net/minecraftforge/fluids/FluidStack.java:
```diff
@@ -84,17 +86,46 @@ public static FluidStack loadFluidStackFromNBT(NBTTagCompound nbt)
+            for (Fluid f : ForgeRegistries.FLUIDS.getValues())
+            {
+                if (f.getRegistryName().getResourcePath().equals(fluidName));
+                {
```
wouldn't a FluidDictionary lookup make more sense here?

Madsthunder replied:
I considered that, but it wouldn't make sense to register EVERY Fluid in `FluidDictionary`. Besides, if a modder is making their fluids backwards-compatible, it would make more sense to just lookup a Fluid using `ForgeRegistries.FLUIDS`, since a Fluid with the name forestry:honey in 1.11 MUST have the name forestry:honey in 1.12 and later. So, as it has worked in 1.11, the Fluid used for the `FluidStack` would be the first Fluid registered through the `FluidRegistry`, or in 1.12, the first Fluid registered in `ForgeRegistries.FLUIDS`.

HenryLoenwind replied:
Doesn't this handle the transition from "old" registry fluids to "new" dictionary fluids? (If not, ignore me.) I would guess that 99% of fluid dictionary names will be the same as the old registry names (at least for those fluids that will be registered there). I wouldn't say the same thing about the rsl names, because those will then be mod internal names---many mods may take this opportunity to use names that fit their naming schema (as I plan to do with our liquid redstone/glowstone/enderpearls).

Madsthunder replied:
Would it be alright to add this before it checks the registry?
`if (FluidDictionary.nameExists(fluidName))
       return FluidDictionary.getFirstFluid(fluidName));`

mezz commented on src/main/java/net/minecraftforge/fluids/FluidDictionary.java:
```diff
@@ -0,0 +1,114 @@
+    
+    public static void registerFluid(Fluid fluid, String name)
+    {
+    	Preconditions.checkArgument(ForgeRegistries.FLUIDS.containsValue(fluid), "Cannot register FluidDictionary entry for unregistered fluid %s", fluid);
```
there is a tab here

mezz commented on src/main/java/net/minecraftforge/fluids/FluidRegistry.java:
```diff
@@ -281,14 +110,8 @@ public static boolean isUniversalBucketEnabled()
-            registerFluid(fluid);
-        }
+    	Preconditions.checkNotNull(fluid, "Fluid cannot be null");
+    	Preconditions.checkArgument(ForgeRegistries.FLUIDS.containsValue(fluid), "Cannot register bucket for unregistered Fluid %s", fluid);
```
there are tabs on these two lines

mezz reviewed (changes requested):
Looking good. Please fix the lines Henry pointed out and the few indentation problems I found.

Madsthunder committed:
Apply some suggestions (2)

Madsthunder committed:
Apply some suggestions (2)

Madsthunder committed:
Apply some suggestions (2)

Madsthunder committed:
Apply some suggestions (2)

Madsthunder committed:
Apply some suggestions (3)

Madsthunder committed:
Apply some suggestions (3)

Madsthunder committed:
Apply some suggestions (3)

Madsthunder commented:
@mezz @HenryLoenwind Removed some random tabs, fixed some grammar errors, added `Locale.ENGLISH` to a couple of `String#toLowerCase` calls, and removed the old `Fluid` constructor.

mezz commented on src/main/java/net/minecraftforge/common/ForgeModContainer.java:
```diff
@@ -492,15 +493,23 @@ public NBTTagCompound getDataForWriting(SaveHandler handler, WorldInfo info)
+            snapshot.entries.put(PersistentRegistryManager.FLUIDS, entry);
+            NBTTagList fluidList = tag.getTagList("DefaultFluidList", 8);
+            for (int id = 0; id < fluidList.tagCount(); id++)
+                entry.ids.put(new ResourceLocation(fluidList.getStringTagAt(id).toLowerCase()), id);
```
toLowerCase is missing Locale

mezz commented:
Looking pretty good. 
I'll take a closer look when 1.12 is coming up.
In the meantime, more people should look it over and comment.

Madsthunder committed:
Apply some suggestions (4)

Madsthunder committed:
Apply some suggestions (4)

Madsthunder committed:
Apply some suggestions (4)

Madsthunder committed:
Add missing import in ForgeModContainer

Madsthunder committed:
Add missing import in ForgeModContainer

Madsthunder committed:
Add missing import in ForgeModContainer

liach commented:
Ping @KingLemming since he designed fluid system originally.

williewillus commented:
If we're doing that then @cpw too for persistent registry :P

KingLemming commented:
At first blush, this basically seems like it's just begging for Fluid duplication. I don't know if this is the right call.

I'll look over it in more detail, but my gut reaction is that we're going to see loads more duplicate fluids and their respective blocks, which is going to absolutely chew through block IDs.

mezz committed:
Fix Universal bucket handling for Fluids with NBT (Fix #3757)

mezz committed:
Update mcp mappings to 20170611

mezz committed:
Clean up ShapedOreRecipe and ShapelessOreRecipe

Madsthunder committed:
Fixed the bootstrap patch (2)

Madsthunder committed:
Fixed the bootstrap patch (2)

Madsthunder committed:
Merge branch 'fml-controlled-fluids' of https://github.com/Madsthunder/MinecraftForge into fml-controlled-fluids

Madsthunder committed:
Merge branch '1.12.x' into fml-controlled-fluids

tterrag1098 committed:
Fix exception in ShapedOreRecipe.checkMatch for recipes that don't fill entire crafting grid (#3995)

marvin-roesch committed:
Make registries persist by default, fixes #3989 (#3998)

LexManos committed:
Fix OreIngredient comparing items in reverse. Closes #3999

LexManos committed:
Fix NBT ItemStack support in json loader. Closes #4002

LexManos committed:
Filter any recipe starting with _, Such names are reserved for metadata.
Actually implement conditions for recipes.

cpw committed:
Patch a small bug in SPacketLoginSuccess that lets us emulate the network
handshake. This will be very useful as I rebuild the networking.

shadowfacts committed:
Allow condition factories to be registered from JSON (#4015)

mezz committed:
Update MCP mappings to 20170617

mezz committed:
Add missing Override annotations

mezz committed:
Add timeout to TestNetworkHandshake setup

mezz committed:
Disable TestNetworkHandshake to make Travis work again

mezz committed:
Log info on Travis tests

mezz committed:
Fix itemstack comparison with null

LexManos committed:
First pass of registry rewrite.
Tests do not run yet.
Things of note:
Removed the idea of substitutions, just register multiple items with the same name they will override each other like a normal map.
Decoupled Forge registries from vanilla classes. They now use bouncer classes. MODDERS SHOULD NEVER USE THESE
Introduced more stringent registry locking. As it sits things are only allowed to be registered during the RegistryEvent.Registry phases!
This is to force modders to split up their registrations, and pave the way for calling these functions multiple times during a single MC lifecycle.
ObjectHolder, Missing Mappings, etc... *should* now work for all registry types, nothing should be special cased to Blocks/Items anymore.
Added optional generic dummy factory to registries, to allow registry creators the ability to dummy more then just Blocks.
Re-worked the ItemStack transformer and applied it to ItemBlock and ItemBlockSpecial. Allowing us to cleanup reflective hacks in Block callbacks.
Registry onAdd callback is now only fired on the ACTIVE registry, fixing any issues of temp registries overriding active objects.

bs2609 committed:
Remove unused patch to ItemRecord. Closes #2686. (#4035)

lumien231 committed:
[1.12] Fixes Exception when getting modlist for a vanilla client (#4028)

williewillus committed:
Fix shapeless recipes not being able to use NBT on output, and fixed issue with null groups.

LexManos committed:
Fix loading recipe jsons from jar files.

mezz committed:
Fix patches in Chunk and Explosion (#4036)

TechnicianLP committed:
IRecipeConditions forge:and and forge:false (#4017)

LexManos committed:
More work, game loads and runs now.
Moved Registry events to directly AFTER PreInit instead of before.
This allows modders to register handlers for it without @EventBusSubscriber.
It also allows you to register custom things needed before the construction of blocks or items such as Fluids.
TODO: Move Fluids to a real registry.
@ObjectHolder can now be used on private fields.
Reworked FMLModIdMappingEvent to include data for all registries.
Tile Entities are now registrable.

mezz committed:
Fix Chunk  patch (#4036)

LexManos committed:
Fixup some registry tests, dummies are now persisted across syncing.

bs2609 committed:
Update NoteBlockEvent with new instruments. Resolves #4018. (#4019)

LexManos committed:
Fix recipe names in packaged jars. Closes #4039

mezz committed:
Fix ShapedOreRecipe matching. #4038

LexManos committed:
Disable substitution test, substitutions are not managed the way they used to be.
We do not care about substitutions, People who replace entries should make them compatible with the existing item.

diesieben07 committed:
Clean up Logging (#3953)

# Conflicts:
#	patches/minecraft/net/minecraft/client/renderer/texture/Stitcher.java.patch
#	patches/minecraft/net/minecraft/client/renderer/texture/TextureMap.java.patch
#	patches/minecraft/net/minecraft/entity/EntityList.java.patch
#	patches/minecraft/net/minecraft/tileentity/TileEntity.java.patch
#	src/main/java/net/minecraftforge/oredict/OreDictionary.java

LexManos committed:
Implement override tracking and syncing.
Overrides MUST be registered within a tracked event so we know what mod it came from.
This will allow servers/saves to select which mod 'wins' and becomes to active entry.
This should also mean that when connecting to a vanilla server things will revert.

LexManos committed:
Fix conflicts

Madsthunder committed:
Merge branch 'fml-controlled-fluids' of https://github.com/Madsthunder/MinecraftForge into fml-controlled-fluids

Merge Conflicts (1)

Madsthunder commented:
Opening a new pr because I somehow screwed up merge conflicts
